### PR TITLE
Refactor deployment

### PR DESCRIPTION
**Pull request by crazy-max**
Closes #14 

* Rename `buildkit-helper` target to `buildkit`
* Merge `publish` job to `build` job in `ci` workflow (now handles with bake-action)
* Review deployment and now using `workflow_dispatch` event with:
  * `qemu_version` input to specify the current QEMU version used
  * `release_number` input to specify the current release number of binfmt
  * `latest` input to create latest tag or not
  * `dry-run` input to release without push
* Change tagging strategy: `<qemu_version>-<release_number>`
* Create following Docker tags on release:
  * `<qemu_version>-<release_number>`, `latest`
  * `buildkit-<qemu_version>-<release_number>`, `buildkit-latest`
* And also create Git tag `<qemu_version>-<release_number>`

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      -
+        name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+        with:
+          image: ${{ env.REPO_SLUG }}:master
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Create Release
```
Comment (tonistiigi) [inline .github/workflows/release.yml:48]: shouldn't this be after bake?

Comment (crazy-max) [reply .github/workflows/release.yml]: Yes

**Comment by crazy-max** [comment]
@tonistiigi I've changed quite the current behavior:

* Replace `qemu_version` input with `target` to choose between mainline or buildkit bake target.
* Remove `release_number` input and use [`GITHUB_RUN_NUMBER`](https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables) introduced recently.
* I've renamed `QEMU_VERSION` with `QEMU_REF` and resolves `QEMU_VERSION` through [VERSION file](https://github.com/qemu/qemu/blob/v5.2.0/VERSION) which seems more accurate.
* GitHub release body is generated with QEMU version and sha based on QEMU_REF.

File: Dockerfile [hunk 2]
```diff
@@ -75,9 +76,10 @@ FROM base-$TARGETARCH$TARGETVARIANT AS base
 
 FROM base AS build
 ARG TARGETPLATFORM
-ARG QEMU_VERSION
+ARG QEMU_REF
+ARG QEMU_REPO
 RUN --mount=target=.,from=src,src=/src/qemu,rw --mount=target=./install-scripts,src=scripts \
-  TARGETPLATFORM=${TARGETPLATFORM} configure_qemu.sh && \
+  TARGETPLATFORM=${TARGETPLATFORM} QEMU_VERSION=v$(curl -L ${QEMU_REPO}/raw/${QEMU_REF}/VERSION) configure_qemu.sh && \
```
Comment (tonistiigi) [inline Dockerfile:82]: why github specific curl and not just open local file?

Comment (crazy-max) [reply Dockerfile:82]: right in this case we can use local file

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    inputs:
+      target:
+        description: 'Bake target'
+        required: true
+        default: 'mainline'
+      latest:
+        description: 'Create latest tag'
+        required: true
+        default: 'false'
+      dry-run:
```
Comment (tonistiigi) [inline .github/workflows/release.yml:14]: would be good to override qemu in here. For test releases, rollbacks etc.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        uses: actions/checkout@v2
+      -
+        name: Prepare
+        run: |
+          QEMU_REPO=$(docker buildx bake --print ${{ github.event.inputs.target }} | jq --raw-output '.. | .args?.QEMU_REPO | select(.)')
+          QEMU_REF=$(docker buildx bake --print ${{ github.event.inputs.target }} | jq --raw-output '.. | .args?.QEMU_VERSION | select(.)')
+          QEMU_VERSION=v$(curl -L ${QEMU_REPO}/raw/${QEMU_REF}/VERSION)
+          QEMU_SHA=$(git ls-remote --sort='version:refname' ${QEMU_REPO} | grep ${QEMU_REF} | cut -f 1 | head -1)
+          TAG=${QEMU_VERSION}-${{ env.GITHUB_RUN_NUMBER }}
+          if [ "${{ github.event.inputs.target }}" = "mainline" ]; then
```
Comment (tonistiigi) [inline .github/workflows/release.yml:37]: this is opposite I think

Comment (crazy-max) [reply .github/workflows/release.yml:43]: yes :)

File: .github/workflows/release.yml (hunk 3 above)
Comment (crazy-max) [reply .github/workflows/release.yml:17]: I had to use an HCL function to handle default values. Let me know if it looks good to you: https://github.com/tonistiigi/binfmt/pull/24/commits/7bf70fe0e75c5623d221aafacdfb23d0b4de87e1

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v2
+      -
+        name: Prepare
+        run: |
+          QEMU_REPO=$(docker buildx bake --print ${{ github.event.inputs.target }} | jq --raw-output '.. | .args?.QEMU_REPO | select(.)')
```
Comment (tonistiigi) [inline .github/workflows/release.yml:38]: why do you parse out the values form bake? instead of just setting the value with a condition like you do already.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          if [ "${{ github.event.inputs.qemu_repo }}" != "" ]; then
+            QEMU_REPO=${{ github.event.inputs.qemu_repo }}
+          fi
+          QEMU_VERSION=$(docker buildx bake --print ${{ github.event.inputs.target }} | jq --raw-output '.. | .args?.QEMU_VERSION | select(.)')
+          if [ "${{ github.event.inputs.qemu_version }}" != "" ]; then
+            QEMU_VERSION=${{ github.event.inputs.qemu_version }}
+          fi
+          QEMU_SHA=$(git ls-remote --sort='version:refname' ${QEMU_REPO} | grep ${QEMU_VERSION} | cut -f 1 | head -1)
```
Comment (tonistiigi) [inline .github/workflows/release.yml:46]: why is this needed?

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,98 @@
+name: release
+
+on:
+  workflow_dispatch:
+    inputs:
+      qemu_repo:
+        description: 'QEMU repo (e.g. https://github.com/qemu/qemu)'
+        required: false
+      qemu_version:
```
Comment (tonistiigi) [inline .github/workflows/release.yml:9]: how would this do a current buildkit release where source is set with a sha?

File: docker-bake.hcl [hunk 8]
```diff
@@ -21,15 +47,24 @@ target "binaries" {
...
+    "linux/386",
+    "linux/mips64le"
+  ]
 }
 
 target "mainline" {
   args = {
-    QEMU_REPO = "https://github.com/qemu/qemu"
-    QEMU_VERSION = "v5.2.0"
+    QEMU_REPO = trimspace(getdef("${QEMU_REPO}", "https://github.com/qemu/qemu"))
```
Comment (tonistiigi) [inline docker-bake.hcl:64]: This looks super hacky. I guess it is because of the shared var that is updated in build master. But without it I would prefer `--set` then.

File: .github/workflows/release.yml (hunk 5 above)
Comment (crazy-max) [reply .github/workflows/release.yml:38]: I think it's more appropriate to use defaults from the bake def if inputs are empty. This allows us not to duplicate these values in the workflow and also to extract the right values according to the selected target.

File: .github/workflows/release.yml (hunk 6 above)
Comment (crazy-max) [reply .github/workflows/release.yml:46]: Yes it's not needed anymore since I've removed the `qemu_ref` input.

File: .github/workflows/release.yml (hunk 7 above)
Comment (crazy-max) [reply .github/workflows/release.yml:9]: `target` input handle the bake target to run. If you set `qemu_version` with a sha the git/docker tag will be the sha commit (maybe we could use a short tag if a long one is provided). That's why I thought it would be better to extract the internal [version available in qemu repo](https://github.com/qemu/qemu/blob/master/VERSION) to create proper docker/git tag.

File: docker-bake.hcl (hunk 8 above)
Comment (crazy-max) [reply docker-bake.hcl:64]: > This looks super hacky. I guess it is because of the shared var that is updated in build master.

Yes I would have preferred to use shared var available on current master branch of buildx.

> But without it I would prefer `--set` then.

Yes I will make a change to use `--set`. That's way better.

File: .github/workflows/release.yml (hunk 5 above)
Comment (tonistiigi) [reply .github/workflows/release.yml:38]: I don't think this is pattern we should be suggesting to any bake users. The logic is that defaults are defined in file and you can override them if needed. Not that you need to use jq to parse out the values from the file and then substitute them with your own values if needed.

File: .github/workflows/release.yml (hunk 7 above)
Comment (tonistiigi) [reply .github/workflows/release.yml:9]: Why not just input for the commit tag. That default to source ref (or to not tagging at all)

File: .github/workflows/release.yml (hunk 5 above)
Comment (crazy-max) [reply .github/workflows/release.yml:38]: I completely agree with that but here we have a rough case where we need to create a GitHub Release in the workflow with its git tag (which is `$QEMU_VERSION-$GITHUB_RUN_NUMBER`) so I have to grab default QEMU_VERSION from bake definition if none is defined as input in the workflow dispatch event.

File: .github/workflows/release.yml (hunk 7 above)
Comment (crazy-max) [reply .github/workflows/release.yml:9]: I guess we could find the most recent tag in the tree from the sha and prepend its value. e.g. `v5.2.0-be25039`, WDYT?

Comment (tonistiigi) [reply .github/workflows/release.yml:9]: I don't think we need lots of magic autodetection. This will be used infrequently by a small number of people. They can give the correct inputs, just needs to be a way to enter it.

**Review by tonistiigi — APPROVED** [review]
